blockchain/difficulty.go

```diff
@@ -743,19 +748,277 @@ func (b *BlockChain) calcNextRequiredStakeDifficulty(curNode *blockNode) (int64,
+	}
+
+	// Get the pool size and number of tickets that were immature at the
+	// previous retarget interval
+	//
+	// NOTE: Since the stake difficulty must be calculated based on existing
+	// blocks, it is always calcuated for the block after a given block, so
+	// the information for the previous retarget interval must be retrieved
+	// relative to the block just before it to coincide with how it was
+	// originally calculated.
+	var prevPoolSize int64
+	prevRetargetHeight := nextHeight - intervalSize - 1
+	prevRetargetNode, err := b.ancestorNode(curNode, prevRetargetHeight)
+	if err != nil {
+		return 0, err
+	}
+	if prevRetargetNode != nil {
+		prevPoolSize = int64(prevRetargetNode.header.PoolSize)
+	}
+	ticketMaturity := int64(b.chainParams.TicketMaturity)
+	prevImmatureTickets, err := b.sumPurchasedTickets(prevRetargetNode,
+		ticketMaturity)
+	if err != nil {
+		return 0, err
+	}
+
+	// Return the existing ticket price for the first few intervals to avoid
+	// division by zero and encourage initial pool population.
+	prevPoolSizeAll := prevPoolSize + prevImmatureTickets
+	if prevPoolSizeAll == 0 {
+		return curDiff, nil
+	}
+
+	// Count the number of currently immature tickets.
+	immatureTickets, err := b.sumPurchasedTickets(curNode, ticketMaturity)
+	if err != nil {
+		return 0, err
+	}
+
+	// Calculate and return the final next required difficulty.
+	curPoolSizeAll := int64(curNode.header.PoolSize) + immatureTickets
+	return calcNextStakeDiffV2(b.chainParams, nextHeight, curDiff,
+		prevPoolSizeAll, curPoolSizeAll), nil
+}
+
+// sdiffAlgoDeploymentVersion returns the deployment vesion for the stake
+// difficulty algorithm change as defined in DCP0001 for the provided network.
+//
+// This function is safe for concurrent access.
+func sdiffAlgoDeploymentVersion(network wire.CurrencyNet) uint32 {
+	if network != wire.MainNet {
+		return 5
+	}
+	return 4
+}
+
+// calcNextRequiredStakeDifficulty calculates the required stake difficulty for
+// the block after the passed previous block node based on the active stake
+// difficulty retarget rules.
+//
+// This function differs from the exported CalcNextRequiredDifficulty in that
+// the exported version uses the current best chain as the previous block node
+// while this function accepts any block node.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcNextRequiredStakeDifficulty(curNode *blockNode) (int64, error) {
+	// Use the new stake difficulty algorithm if the stake vote for the new
+	// algorithm agenda is active.
+	//
+	// NOTE: The choice field of the return threshold state is not examined
+	// here because there is only one possible choice that can be active
+	// for the agenda, which is yes, so there is no need to check it.
+	deploymentVersion := sdiffAlgoDeploymentVersion(b.chainParams.Net)
+	state, err := b.deploymentState(curNode, deploymentVersion,
+		chaincfg.VoteIDSDiffAlgorithm)
+	if err != nil {
+		return 0, err
```

Comment:
Is this right?  What happens on the other end of this function?  Should this be either a panic or a fallthrough to have some defaults?